### PR TITLE
Use of SIMD computation in CRHMC walk

### PR DESCRIPTION
**Pull request by iakoviid**
Expanding on PR #239 we use the SIMD capabilities of PackedCSparse and present a vectorized CRHMC walk.
This way we can have speed up on medium to large datasets, depending on the systems architecture.
Changes from #239 :
1) Changes in CRHMC walk instead of using `n`-dimensional vectors we use `n * simd_len`-dimensional matrices. In this way we do simd_len parallel CRHMC walks without spawning additional threads.
2) Changes in testing and benchmarking files to test the behavior of the added code.

### CODE REVIEW

**Review by vissarion — APPROVED** [review]
I am OK with merging this PR. Thanks, that is a really cool functionality! 

There is only one issue (or maybe two) with the R tests that are failing in windows / mac. Please fix it and we can proceed with merging.

File: R-proj/tests/testthat/test_sampling.R [hunk 1]
```diff
@@ -14,63 +14,85 @@ runsample <- function(P, name_string, dist){
     res = 1
   }
   return(res)
-  
+
+}
+
+logconcave_sample<- function(P,distribution, n_samples ,n_burns){
```
Comment (papachristoumarios) [inline R-proj/tests/testthat/test_sampling.R:20]: fix coding style

File: examples/crhmc_sampling/common.hpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
+#ifndef COMMON_HPP
```
Comment (papachristoumarios) [inline examples/crhmc_sampling/common.hpp]: Can you give a more descriptive name to the file?

File: examples/crhmc_sampling/data/afiro.mm [hunk 3]
```diff
@@ -0,0 +1,112 @@
+%%MatrixMarket matrix coordinate real general
```
Comment (papachristoumarios) [inline examples/crhmc_sampling/data/afiro.mm:1]: What does MatrixMarket mean?

File: examples/crhmc_sampling/README.md [hunk 4]
```diff
@@ -16,6 +16,6 @@ For example: -DLP_SOLVE=/usr/lib/lpsolve/liblpsolve55.so
```
Comment (papachristoumarios) [inline examples/crhmc_sampling/README.md:16]: Please add plots from the results from the various polytopes.

**Review by papachristoumarios — APPROVED** [review]
Impressive job! I have a few comments, mostly of aesthetical nature, but overall the PR looks good to me!

File: examples/crhmc_sampling/data/afiro.mm (hunk 3 above)
Comment (iakoviid) [reply examples/crhmc_sampling/data/afiro.mm:1]: It is a popular sparse matrix file format: https://math.nist.gov/MatrixMarket/mmio-c.html

File: examples/crhmc_sampling/README.md (hunk 4 above)
Comment (iakoviid) [reply examples/crhmc_sampling/README.md:16]: Done.

File: examples/crhmc_sampling/common.hpp (hunk 2 above)
Comment (iakoviid) [reply examples/crhmc_sampling/common.hpp]: done

File: R-proj/tests/testthat/test_sampling.R (hunk 1 above)
Comment (iakoviid) [reply R-proj/tests/testthat/test_sampling.R:20]: done

File: examples/crhmc_sampling/data/afiro.mm (hunk 3 above)
Comment (vissarion) [reply examples/crhmc_sampling/data/afiro.mm:1]: Great, could you please add this sentence as a comment?

**Comment by vissarion** [comment]
@iakoviid there are some issues still in CI tests (most probably new). There is one SEGFAULT with clang-12 and an error in R tests in windows (where some psrf is NaN). Could you please check?

**Review by TolisChal — APPROVED** [review]

**Comment by TolisChal** [comment]
I'm closing this PR since it is included in https://github.com/GeomScale/volesti/pull/286

**Pull request closed without merging**